### PR TITLE
Opaque

### PR DESCRIPTION
**Pull request by jumormt**
add support for opaque pointers (tested with LLVM-14 with getptrelement and getpointerto APIs removed).

### CODE REVIEW

File: svf-llvm/include/SVF-LLVM/LLVMUtil.h [hunk 1]
```diff
@@ -100,9 +100,11 @@ static inline Type* getPtrElementType(const PointerType* pty)
 {
 #if (LLVM_VERSION_MAJOR < 14)
     return pty->getPointerElementType();
-#else
+#elif (LLVM_VERSION_MAJOR < 17)
```
Comment (yuleisui) [inline svf-llvm/include/SVF-LLVM/LLVMUtil.h:103]: I remember it is version 16?

File: svf-llvm/include/SVF-LLVM/ObjTypeInference.h [hunk 2]
```diff
@@ -105,6 +109,14 @@ class ObjTypeInference
...
     u32_t getArgPosInCall(const CallBase *callBase, const Value *arg);
 
+public:
+    /// get or infer the name(s) of thisptr
+    Set<std::string> &inferThisPtrClassName(const Value *thisPtr);
+
+protected:
+
+    /// Backward collect all possible cpp constructors starting from a value
+    Set<const Value *> findCPPSources(const Value *startValue);
```
Comment (yuleisui) [inline svf-llvm/include/SVF-LLVM/ObjTypeInference.h:119]: FindCPPConstor?

File: svf-llvm/include/SVF-LLVM/LLVMUtil.h (hunk 1 above)
Comment (jumormt) [reply svf-llvm/include/SVF-LLVM/LLVMUtil.h:103]: 16 still supports typed pointers (https://llvm.org/docs/OpaquePointers.html#version-support)

File: svf-llvm/include/SVF-LLVM/ObjTypeInference.h (hunk 2 above)
Comment (jumormt) [reply svf-llvm/include/SVF-LLVM/ObjTypeInference.h:120]: sorry for the confusion, sources can be heap allocation or self-inference functions (constructors/destructors or template functions). I've updated the comments.

File: svf-llvm/include/SVF-LLVM/CppUtil.h [hunk 3]
```diff
@@ -129,6 +129,39 @@ bool VCallInCtorOrDtor(const CallBase* cs);
...
+Set<std::string> extractClassNameInTemplate(const std::string &oname);
+
+/// CPP source used to infer c++ class name
+/// sources can be heap allocation or self-inference functions (constructors/destructors or template functions)
+bool isCPPSource(const Value *val);
+
+/// whether fooName matches the mangler label
+bool matchManglerLabel(const std::string &fooName, const std::string &label);
+
+/// whether foo is a cpp self-inference function
```
Comment (yuleisui) [inline svf-llvm/include/SVF-LLVM/CppUtil.h:145]: What does this mean? Can be self-inferred?

Comment (jumormt) [reply svf-llvm/include/SVF-LLVM/CppUtil.h:145]: functions that we can get the class name from the function name, like constructors (e.g., A::A()), template functions (e.g., std::deque<A const*, std::allocator<A const*> >). Any suggestions on better naming?

**Pull request merged**